lib/dependencies.sh

```diff
@@ -29,6 +29,23 @@ run_if_present() {
   fi
 }
 
+run_build_script() {
+  local has_build_script=$(read_json "$BUILD_DIR/package.json" ".scripts.build")
+  local has_heroku_build_script=$(read_json "$BUILD_DIR/package.json" ".scripts[\"heroku-postbuild\"]")
+
+  if [[ -n "$has_heroku_build_script" ]] && [[ -n "$has_build_script" ]]; then
+    echo "Detected both 'build' and 'heroku-postbuild' scripts"
+    mcount "scripts.heroku-postbuild-and-build"
+    run_if_present 'heroku-postbuild'
+  elif [[ -n "$has_heroku_build_script" ]]; then
+    mcount "scripts.heroku-postbuild"
+    run_if_present 'heroku-postbuild'
+  elif [[ -n "$has_build_script" ]]; then
```

Comment:
If users have a `postinstall` scripts entry of say `yarn build`, I think this will mean their build will be run twice. 

In such cases (and variations on it, such as the `prepare`, `install`, ... lifecycle scripts), is the expectation that users will need to opt-out of the buildpack change or else stop using `postinstall`? (It wasn't clear whether the "warn users" etc in #583 was talking about only the "has build script but no others" case, or this one too.)

Reply:
In the case that `postinstall` is just running `run build`, it's expected that the user should stop using `postinstall`. Most of them have only added it because it's what Heroku has instructed them to do in order to run builds on Heroku, and I regret that decision.

In the case that the `build` script is not meant to be run on Heroku, adding a no-op `heroku-postbuild` script would accomplish that.

I have an `npx migrate-these-build-scripts` that can walk a user through this with one command, and will be adding warnings allowing users to opt-in to the change before it's pushed. We'll also be emailing paid users who will be affected.

Reply:
🤔 I can't avoid running `postinstall` since `npm` and `yarn` run that. I could tell `npm` not to run scripts at all, but it's unlikely that I'd be able to reproduce what it does.

I could skip running the `build` script for those that call it in `postinstall` during an adjustment period. Or skip it if `postinstall` is defined. It's made a little more complicated because many `postinstall` scripts look like this:

`npm rebuild node-sass && yarn run build` or
`sh tasks/postinstall.sh`

or they run other scripts that might run `build`.

Reply:
There are also thousands of apps that use both `postinstall` and `heroku-postbuild`, likely so that `postinstall` will run while they're developing locally. 

I'm not sure I want to try to outsmart our users this way.